crates/dbsp/src/operator/communication/exchange.rs

```diff
@@ -31,7 +60,188 @@ use std::{
+                    } else {
+                        local_client.clone()
+                    };
+                    (host.workers.clone(), client)
+                })
+                .collect(),
+        })
+    }
+}
+
+impl Index<usize> for Clients {
+    type Output = ExchangeServiceClient;
+    fn index(&self, worker: usize) -> &ExchangeServiceClient {
+        for (workers, host) in self.0.iter() {
+            if workers.contains(&worker) {
+                return host;
+            }
+        }
+        unreachable!();
+    }
+}
+
+struct Inner {
+    tokio: TokioHandle,
+    exchange_id: ExchangeId,
+    /// The number of communicating peers.
+    npeers: usize,
+    clients: Arc<Clients>,
+    sender_notifies: Vec<Notify>,
+    /// Counts the number of messages yet to be received in the current round of
+    /// communication per receiver.  The receiver must wait until it has all
+    /// `npeers` messages before reading all of them from mailboxes in one
+    /// pass.
+    receiver_counters: Vec<AtomicUsize>,
+    /// Callback invoked when all `npeers` messages are ready for a receiver.
+    receiver_callbacks: Vec<OnceCell<Box<dyn Fn() + Send + Sync>>>,
+    /// For each sender, whether the sends from the previous round have
+    /// completed, so that the next round can begin.
+    ready_to_send: Vec<AtomicBool>,
+    /// Callback invoked when all `npeers` mailboxes are available.
+    sender_callbacks: Vec<OnceCell<Box<dyn Fn() + Send + Sync>>>,
+    deliver: Box<dyn Fn(Vec<u8>, usize, usize) + Send + Sync + 'static>,
+}
+
+impl Inner {
+    fn new(
+        runtime: &Runtime,
+        tokio: TokioHandle,
+        exchange_id: ExchangeId,
+        deliver: impl Fn(Vec<u8>, usize, usize) + Send + Sync + 'static,
+        clients: Arc<Clients>,
+    ) -> Inner {
+        let _guard = tokio.enter();
+
+        let npeers = runtime.num_workers();
+
+        let receiver_counters: Vec<_> = (0..npeers).map(|_| AtomicUsize::new(0)).collect();
+
+        let receiver_callbacks: Vec<_> = (0..npeers).map(|_| OnceCell::new()).collect();
+
+        let sender_notifies: Vec<_> = (0..npeers * npeers).map(|_| Notify::new()).collect();
+
+        Self {
+            tokio: tokio.clone(),
+            exchange_id,
+            npeers,
+            clients,
+            receiver_counters,
+            receiver_callbacks,
+            sender_notifies,
+            ready_to_send: (0..npeers).map(|_| AtomicBool::new(true)).collect(),
+            sender_callbacks: (0..npeers).map(|_| OnceCell::new()).collect(),
+            deliver: Box::new(deliver),
+        }
+    }
+
+    async fn received(self: &Arc<Self>, data: Vec<u8>, sender: usize, receiver: usize) {
+        let index = sender * self.npeers + receiver;
+
+        (self.deliver)(data, sender, receiver);
+
+        let old_counter = self.receiver_counters[receiver].fetch_add(1, Ordering::AcqRel);
+        if old_counter >= self.npeers - 1 {
+            // This can be a spurious callback.
```

Comment:
I am not sure what that means.

Reply:
This was an existing comment already in the code that I just moved around. I'll try to improve it.